esp-hal/src/sdio/esp32c6.rs

```diff
@@ -0,0 +1,106 @@
+use esp32c6::{SLC, SLCHOST};
+
+use embedded_hal::mmc::{CardMode, CardType, FifoStatus, MmcOps, Reset};
+use embedded_hal::mmc::command::MmcCommand;
+use embedded_hal::mmc::response::MmcResponse;
+use embedded_hal::mmc::tuning::{TuningMode, TuningWidth};
+
+use super::result::{Error, Result};
+
+/// Represents the SDIO 2.0 peripheral for the ESP32C6 microcontroller.
+pub struct Sdio {
+    slc: SLC,
+    slchost: SLCHOST,
+}
+
+impl Sdio {
+    /// Creates a new [Sdio].
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use esp32c6::Peripherals;
+    /// use esp_hal::sdio::Sdio;
+    ///
+    /// let dp = Peripheral::take().unwrap();
+    /// let _sdio = Sdio::new(dp.slc, dp.slchost);
+    /// ```
+    pub const fn new(slc: SLC, slchost: SLCHOST) -> Self {
+        Self { slc, slchost }
+    }
+
+    /// Releases the inner SDIO peripherals back to the caller.
+    pub const fn release(self) -> (SLC, SLCHOST) {
```

Comment:
FYI this pattern goes against our guidelines in esp-hal: https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md.

We use peripheral singletons with lifetime parameters, you can "borrow" (our method is called `reborrow`, but now I'm wondering if `borrow` makes more sense cc: @bugadani) a peripheral to allow drivers to automatically "release" the peripheral singleton again once the driver has been dropped.

Reply:
I almost completely refactored using your advice, and the developer guidelines. Also took a look at some other peripheral implementations, and hopefully got closer to the intended style of `esp-hal`.